File: 8. String to Integer (atoi)/8. String to Integer (atoi).py
class Solution:
    def myAtoi(self, s: str) -> int:
        INT_MAX : int = 2**31 - 1
        INT_MIN : int = -2**31

        i : int = 0
        n : int = len(s)
        result : int = 0
        sign : int = 1

        # Step 1: Skip leading whitespaces
        while i < n and s[i] == ' ':
            i += 1

        # Step 2: Check for sign
        if i < n and (s[i] == '+' or s[i] == '-'):
            sign = -1 if s[i] == '-' else 1
            i += 1

        # Step 3: Convert digits and stop on first non-digit
        while i < n and s[i].isdigit():
            digit : int = int(s[i])

            # Step 4: Check for overflow/underflow
            if result > (INT_MAX - digit) // 10:
                return INT_MAX if sign == 1 else INT_MIN

            result = result * 10 + digit
            i += 1

        return sign * result

File: 8. String to Integer (atoi)/8. String to Integer (atoi).c
#include <string.h>
#include <ctype.h>
#include <limits.h>

int myAtoi(char* s) {
    int i = 0;
    int result = 0;
    int sign = 1;
    int n = strlen(s);

    // Step 1: Skip leading whitespaces
    while (i < n && s[i] == ' ') {
        i++;
    }

    // Step 2: Check the sign
    if (i < n && (s[i] == '+' || s[i] == '-')) {
        sign = (s[i] == '-') ? -1 : 1;
        i++;
    }

    // Step 3: Convert digits and stop on the first non-digit
    while (i < n && isdigit((unsigned char)s[i])) {
        int digit = s[i] - '0';

        // Step 4: Check for overflow/underflow
        if (result > (INT_MAX - digit) / 10) {
            return sign == 1 ? INT_MAX : INT_MIN;
        }

        result = result * 10 + digit;
        i++;
    }

    return result * sign;
}